src/providers/appium.ts

```diff
@@ -0,0 +1,129 @@
+import { ChildProcess, spawn, exec } from "child_process";
+import path from "path";
+
+export function startAppiumServer(provider: string): Promise<ChildProcess> {
+  let isEmulatorStarted = false;
+  return new Promise((resolve, reject) => {
+    const appiumProcess = spawn("npx", ["appium"], {
+      stdio: "pipe",
+    });
+
+    appiumProcess.stdout.on("data", async (data: Buffer) => {
+      const output = data.toString();
+      console.log(output);
+
+      if (output.includes("Could not find online devices")) {
+        if (!isEmulatorStarted && provider == "emulator") {
+          isEmulatorStarted = true;
+          await startAndroidEmulator();
+        }
+        resolve(appiumProcess);
+      }
+
+      if (output.includes("Appium REST http interface listener started")) {
+        console.log("Appium server is up and running.");
+        resolve(appiumProcess);
+      }
+    });
+
+    appiumProcess.on("error", (error) => {
+      console.log("Error starting Appium server:", error);
+      reject(error);
+    });
+
+    process.on("exit", () => {
+      console.log("Main process exiting. Killing Appium server...");
+      appiumProcess.kill();
+    });
+
+    appiumProcess.on("close", (code: number) => {
+      console.log(`Appium server exited with code ${code}`);
+    });
+  });
+}
+
+async function startAndroidEmulator(deviceName?: string): Promise<void> {
+  return new Promise((resolve, reject) => {
+    const androidHome = process.env.ANDROID_HOME;
+
+    if (!androidHome) {
+      return reject("ANDROID_HOME environment variable is not set.");
+    }
+
+    const emulatorPath = path.join(androidHome, "emulator", "emulator");
+
+    //@ts-ignore
+    exec(`${emulatorPath} -list-avds`, (error, stdout, stderr) => {
+      if (error) {
+        return reject(`Error fetching emulator list: ${error.message}`);
+      }
+      if (stderr) {
+        console.error(`stderr: ${stderr}`);
+      }
+
+      const devices = stdout.trim().split("\n");
+      console.log(`Available emulators: ${devices.join(", ")}\n\n`);
```

Comment:
strip the info log, split and get the devices and then put log if no emulators found and if found then the list of devices and the device picked for the test